blog/posts/dune-universal-libraries-preview.md

```diff
@@ -0,0 +1,241 @@
+---
+title: Unlocking Universal Libraries in Dune
+date: 2024-04-10
+author: Antonio Monteiro
+gravatar: 45c2052f50f561b9dc2cae59c777aecd794f57269fa317f9c9c3365c2e00d16f
+twitter: '@_anmonteiro'
+---
+
+I recently shared a [2024 progress
+update](whats-2024-brought-to-melange-so-far) about our work on Melange. In
+that message, I briefly wrote about "universal libraries" in Dune, the ability
+to write a shared OCaml / Melange codebase while varying specific module
+implementations, flags, preprocessing steps, etc. according to the compilation
+target.
+
+I also promised to dive deeper into what "universal libraries" are all about,
+and the new use cases that they unlock in Dune. Keep reading for an in-depth
+look at the history behind this new feature rolling out in Dune 3.16.
+
+---
+
+## The Bird's-eye View
+
+Let's walk backwards from our end-goal: having a shared OCaml / Melange
+codebase that can render React.js components on the server, such that the
```

Comment:
I don't consider universal to only render components, but also work with encoding/decoding JSON, share types and validations.

Reply:
yeah, me too. In this example, however, I'm walking backwards from a concrete use case, not defining universal libraries.